v3_async_wip/tests/test_mqtt_client.py

```diff
@@ -0,0 +1,1368 @@
+        assert mock_paho.connect.call_count == 1
+        assert not connect_task.done()
+        # Attempt disconnect
+        disconnect_task = asyncio.create_task(client.disconnect())
+        await asyncio.sleep(0.1)
+        # Paho disconnect has NOT been called yet and task is still pending
+        assert mock_paho.disconnect.call_count == 0
+        assert not disconnect_task.done()
+
+        # Complete connect
+        rc = 0 if pending_success else 1
+        mock_paho.trigger_connect(rc=rc)
+        await asyncio.sleep(0.1)
+        assert connect_task.done()
+
+        if pending_success:
+            assert client.is_connected()
+            # Disconnect was invoked on Paho and is waiting for completion
+            assert not disconnect_task.done()
+            assert mock_paho.disconnect.call_count == 1
+            # Complete the disconnect
+            mock_paho.trigger_disconnect(rc=0)
+            await disconnect_task
+            assert not client.is_connected()
+        else:
+            assert not client.is_connected()
+            # Disconnect was completed without invoking connect on Paho because it is
+            # already disconnected
+            assert disconnect_task.done()
+            assert mock_paho.disconnect.call_count == 0
+            assert not client.is_connected()
+
+    # NOTE: Disconnect can't fail
+    @pytest.mark.it("Waits for a pending disconnect task to finish before attempting a disconnect")
+    async def test_disconnect_pending_disconnect(self, client, mock_paho):
+        # Require manual completion
+        mock_paho._manual_mode = True
+        assert mock_paho.disconnect.call_count == 0
+        # Client is currently connected
+        client_set_connected(client)
+        assert client.is_connected()
+
+        # Attempt first disconnect
+        disconnect_task1 = asyncio.create_task(client.disconnect())
+        await asyncio.sleep(0.1)
+        # Paho disconnect has been called but task 1 is still pending
+        assert mock_paho.disconnect.call_count == 1
+        assert not disconnect_task1.done()
+        # Attempt second disconnect
+        disconnect_task2 = asyncio.create_task(client.disconnect())
+        await asyncio.sleep(0.1)
+        # Paho disconnect has NOT been called an additional time and task2 is still pending
+        assert mock_paho.disconnect.call_count == 1
+        assert not disconnect_task2.done()
+
+        # Complete first disconnect
+        mock_paho.trigger_disconnect(rc=0)
+        await disconnect_task1
+        assert not client.is_connected()
+
+        # Second disconnect was completed without invoking disconnect on Paho because it is
+        # already disconnected
+        await disconnect_task2
+        assert mock_paho.disconnect.call_count == 1
+        assert not client.is_connected()
+
+
+@pytest.mark.skipif(sys.version_info < (3, 8), reason="AsyncMock not supported by Python 3.7")
```

Comment:
Once I get out of this MQTT layer I'll find a solution for backporting AsyncMocks to 3.7 since we'll need them for testing other layers/modules and we can't drop 3.7 until December 2023